File: Python/cognex_camera.py
# An example of implementing a Cognex camera interface using the WebAPI in Python.
import asyncio
import json
import websockets
import logging
import httpx
import socket
import time

class CogSocket:
    def __init__(self, websocket_uri, root=None):
        self.uri = websocket_uri
        self.root = root
        self.websocket = None
        self.request_id = 0
        self.pending_requests = {}
        self.listeners = {}
        self.space = None
        self.log = None
        self.onopen = None
        self.onerror = None
        self.onclose = None

    # Connect to a URI.
    async def connect(self):
        try:
            self.websocket = await websockets.connect(self.uri)
            if self.onopen:
                self.onopen()
            asyncio.create_task(self._listen())
        except Exception as e:
            if self.onerror:
                self.onerror()
            raise

    # Listen for messages.
    async def _listen(self):
        try:
            async for message in self.websocket:
                await self._handle_message(message)
        except websockets.exceptions.ConnectionClosed:
            if self.onclose:
                self.onclose()

    # Handle message responses.
    async def _handle_message(self, message):
        if self.log:
            self.log(f"Received: {message}")
        try:
            msg = json.loads(message)
            msg_type = msg.get('$type')
            path = msg.get('path')
            if msg_type == 'resp':
                req_id = msg.get('id')
                pending = self.pending_requests.get(req_id)
                if pending:
                    future = pending.get('future')
                    if future:
                        if 'error' in msg:
                            error = Exception(msg.get('body', str(msg['error'])))
                            future.set_exception(error)
                        else:
                            future.set_result(msg.get('body'))
                    del self.pending_requests[req_id]
            elif msg_type == 'event':
                listeners = self.listeners.get(path, [])
                body = msg.get('body')
                if body is not None:
                    args = body if isinstance(body, list) else [body]
                else:
                    args = []
                for listener in listeners:
                    if asyncio.iscoroutinefunction(listener):
                        await listener(*args)
                    else:
                        listener(*args)
        except json.JSONDecodeError:
            pass

    # Create a unique ID.
    def _next_request_id(self):
        self.request_id = (self.request_id + 1) % 0x7FFFFFFF
        return self.request_id

    # Send a request.
    async def _send_request(self, type_, path, body=None):
        req_id = self._next_request_id()
        future = asyncio.Future()
        self.pending_requests[req_id] = {'future': future}
        msg = {
            '$type': type_,
            'id': req_id,
            'path': path
        }
        if body is not None:
            msg['body'] = body
        json_msg = json.dumps(msg, indent=self.space)
        if self.log:
            self.log(f"Send: {json_msg}")
        await self.websocket.send(json_msg)
        return await future

    # Send a GET request.
    async def get(self, path):
        return await self._send_request('get', path)

    # Send a PUT request.
    async def put(self, path, data):
        return await self._send_request('put', path, data)

    # Send a POST request.
    async def post(self, path, data):
        return await self._send_request('post', path, data)

    # Add message listeners.
    async def add_listener(self, path, listener):
        if path not in self.listeners:
            self.listeners[path] = []
            await self._send_request('listen', path)
        self.listeners[path].append(listener)

    # Remove message listeners.
    async def remove_listener(self, path, listener=None):
        if path in self.listeners:
            if listener:
                try:
                    self.listeners[path].remove(listener)
                except ValueError:
                    pass
                if not self.listeners[path]:
                    del self.listeners[path]
                    await self._send_request('unlisten', path)
            else:
                del self.listeners[path]
                await self._send_request('unlisten', path)

    # Close the connection.
    async def close(self):
        if self.websocket:
            await self.websocket.close()
            self.websocket = None

class CognexCamera:
    def __init__(self, ip, port=80, username='admin', password=''):
        self.ip = ip
        self.port = port
        self.username = username
        self.password = password
        self.cogsock = None
        self.session_id = None
        self.keep_alive_task = None
        self.root = 'cam0/hmi'
        self.cells = 'A0:Z100'
        
        # Event subscriber lists
        self.StateChanged = []
        self.ResultsChanged = []
        self.LiveModeChanged = []
        self.JobInfoChanged = []
        self.EditorAttachedChanged = []
        self.JobLoadingChanged = []
        self.SettingsChanged = []
        self.JobLoadFailed = []
        self.JobValidationDone = []
        self.SessionDisposed = []

    async def Connect(self):
        uri = f"ws://{self.ip}:{self.port}/ws"
        self.cogsock = CogSocket(uri)
        #self.cogsock.log = lambda msg: print(f"[CogSocket] {msg}")
        await self.cogsock.connect()
        # open session
        session_info = {'cellNames': [self.cells]}
        self.session_id = await self.cogsock.post(f"{self.root}/openSession", session_info)
        print(f"Session: {self.session_id}")
        # Login
        ok = await self.cogsock.post(f"{self.session_id}/login", [self.username, self.password, False])
        if isinstance(ok, dict) and ok.get('error'):
            raise Exception("Login failed")
        print("Login successful")
        # Add listeners
        await self.cogsock.add_listener(f"{self.root}/stateChanged", self._on_state_changed)
        await self.cogsock.add_listener(f"{self.session_id}/resultChanged", self._on_result_changed)
        await self.cogsock.add_listener(f"{self.root}/liveModeChanged", self._on_liveMode_changed)
        await self.cogsock.add_listener(f"{self.root}/jobChanged", self._on_job_changed)
        await self.cogsock.add_listener(f"{self.root}/editorAttachedChanged", self._on_editorAttached)
        await self.cogsock.add_listener(f"{self.root}/jobLoadingChanged", self._on_jobLoading_changed)
        await self.cogsock.add_listener(f"{self.root}/settingsChanged", self._on_settings_changed)
        await self.cogsock.add_listener(f"{self.root}/jobLoadFailed", self._on_jobLoadFailed_changed)
        await self.cogsock.add_listener(f"{self.root}/jobValidationDone", self._on_jobValidationDone_changed)
        await self.cogsock.add_listener(f"{self.root}/sessionDisposed", self._on_sessionDisposed_changed)
        # ready
        await self.cogsock.post(f"{self.session_id}/ready", "")
        # Start keep alive
        self.keep_alive_task = asyncio.create_task(self._keep_alive())

    async def _keep_alive(self):
        while True:
            await asyncio.sleep(15)
            if self.session_id:
                await self.cogsock.post(f"{self.session_id}/keepAlive", "")
                await self.cogsock.post(f"{self.session_id}/ready", "")

    async def _on_state_changed(self, *args):
        #print("State changed")

        # ONLY callbacks
        for cb in self.StateChanged:
            await cb(*args)

    async def _on_result_changed(self, *args):
        #print("Result changed")

        # ONLY callbacks
        for cb in self.ResultsChanged:
            await cb(*args)
            
    async def _on_liveMode_changed(self, *args):
        #print("Live mode changed")

        # ONLY callbacks
        for cb in self.LiveModeChanged:
            await cb(*args)

    async def _on_job_changed(self, *args):
        #print("Job changed")

        # ONLY callbacks
        for cb in self.JobInfoChanged:
            await cb(*args)

    async def _on_editorAttached(self, *args):
        #print("Editor attached")

        # ONLY callbacks
        for cb in self.EditorAttachedChanged:
            await cb(*args)

    async def _on_jobLoading_changed(self, *args):
        #print("Job loading changed")

        # ONLY callbacks
        for cb in self.JobLoadingChanged:
            await cb(*args)
            
    async def _on_settings_changed(self, *args):
        # print("Settings changed")

        # ONLY callbacks
        for cb in self.SettingsChanged:
            await cb(*args)           
            
    async def _on_jobLoadFailed_changed(self, *args):
        # print("Job load failed")

        # ONLY callbacks
        for cb in self.JobLoadFailed:
            await cb(*args)
  
    async def _on_jobValidationDone_changed(self, *args):
        # print("Job validation done")

        # ONLY callbacks
        for cb in self.JobValidationDone:
            await cb(*args)          
            
    async def _on_sessionDisposed_changed(self, *args):
        # print("Session disposed")

        # ONLY callbacks
        for cb in self.SessionDisposed:
            await cb(*args)

    async def Disconnect(self):
        try:
            # Stop keep-alive first
            if self.keep_alive_task:
                self.keep_alive_task.cancel()
                try:
                    await self.keep_alive_task
                except asyncio.CancelledError:
                    pass

            # Dispose session on Cognex server
            if self.cogsock and self.session_id:
                try:
                    print(f"Disposing session: {self.session_id}")
                    await self.cogsock.post(f"{self.session_id}/dispose", None)
                except Exception as e:
                    print(f"Dispose failed: {e}")

            # Close socket
            if self.cogsock:
                await self.cogsock.close()

        finally:
            self.keep_alive_task = None
            self.session_id = None

    # Camera functions
    async def ManualAcquire(self):
        await self.cogsock.post(f"{self.session_id}/manualTrigger", "")

    async def SetLiveModeAsync(self, enabled: bool):
        await self.cogsock.put(f"{self.session_id}/liveMode", enabled)

    async def ToggleOnlineOffline(self):
        resp = await self.cogsock.get(f"{self.session_id}/softOnline")
        current_online = resp if isinstance(resp, bool) else False
        new_online = not current_online
        await self.cogsock.put(f"{self.session_id}/softOnline", new_online)

    async def QueryCellResults(self, cell):
        resp = await self.cogsock.post(f"{self.session_id}/queryCellResults", [[cell]])
        return json.dumps(resp)

    async def GetCellExpression(self, cell):
        resp = await self.cogsock.post(f"{self.session_id}/getCellExpressions", [cell, True])
        return json.dumps(resp)

    async def SetCellExpression(self, cell, expr):
        await self.cogsock.post(f"{self.session_id}/setCellExpression", [cell, expr])

    async def SetCellValue(self, cell, value):
        await self.cogsock.post(f"{self.session_id}/setCellValue", [cell, value])

    async def ListFiles(self):
        resp = await self.cogsock.post(f"{self.session_id}/listFiles", [])
        return json.dumps(resp)

    async def Info(self):
        resp = await self.cogsock.get(f"{self.root}/info")
        return json.dumps(resp)

    async def FindState(self):
        resp = await self.cogsock.get(f"{self.root}/state")
        return json.dumps(resp)

    async def SetSoftOnlineAsync(self, enabled: bool):
        await self.cogsock.put(f"{self.session_id}/softOnline", enabled)

    async def GetJobInfo(self):
        resp = await self.cogsock.get(f"{self.root}/job")
        return json.dumps(resp)
    
    async def SaveJob(self, job_name):
        await self.cogsock.post(f"{self.session_id}/saveJob", job_name)

    async def LoadJob(self, job_name):
        await self.cogsock.post(f"{self.session_id}/loadJob", job_name)

    async def SendReady(self):
        await self.cogsock.post(f"{self.session_id}/ready", "")

    async def GetSessionIDs(self):
        resp = await self.cogsock.post(f"{self.session_id}/getSessionIDs", "")
        return json.dumps(resp)

    async def JobValidationState(self):
        resp = await self.cogsock.get(f"{self.session_id}/jobValidationState")
        return json.dumps(resp)
    
    async def SystemValidationFlag(self):
        resp = await self.cogsock.get(f"{self.session_id}/systemValidationFlag")
        return json.dumps(resp)
    
    async def RunJobValidation(self):
        await self.cogsock.post(f"{self.session_id}/runJobValidation", "")
        
    async def CancelJobValidation(self):
        await self.cogsock.post(f"{self.session_id}/cancelJobValidation", "")

    async def GetKeepAliveTimeout(self):
        resp = await self.cogsock.get(f"{self.root}/keepAliveTimeout")
        return json.dumps(resp)
    
    async def SetKeepAliveTimeout(self, interval_ms):
        resp = await self.cogsock.put(f"{self.root}/keepAliveTimeout", interval_ms)
        return json.dumps(resp)

    async def LoadImage(self, filename, image_name=None):
        # Read file as bytes
        with open(filename, "rb") as f:
            bytes_data = f.read()

        # Build URL
        url = f"http://{self.ip}:{self.port}/{self.session_id}/loadImage"

        timeout = httpx.Timeout(30.0)

        async with httpx.AsyncClient(timeout=timeout) as client:

            # ByteArrayContent + Content-Type header
            resp = await client.post(url, content=bytes_data, headers={"Content-Type": "image/bmp"})

            # EnsureSuccessStatusCode()
            resp.raise_for_status()

    def NMC(self, nmc, timeout, port, ip, username, password):
        sock = socket.create_connection((ip, port))
        sock.settimeout(timeout)

        # Read initial data
        data = sock.recv(4096)
        # print(repr(data.decode('ascii', errors='ignore')))

        # Login
        sock.sendall((username + "\r\n" + password + "\r\n").encode("ascii"))

        time.sleep(timeout)

        # Read response
        data = sock.recv(4096)
        # print(repr(data.decode('ascii', errors='ignore')))

        # Send command
        sock.sendall((nmc + "\r\n").encode("ascii"))

        time.sleep(timeout)

        # Read response
        response = ""
        while True:
            try:
                data = sock.recv(4096)
                if not data:
                    break
                response += data.decode('ascii', errors='ignore')
            except socket.timeout:
                break
            
        sock.close()   
        
        return response

    # Get the startup job
    async def GetStartupJob(self):
        resp = await self.cogsock.get(f"{self.session_id}/startupJob")
        return json.dumps(resp)
    
    # Set the startup job
    async def SetStartupJob(self, job_name):
        await self.cogsock.put(f"{self.session_id}/startupJob", job_name)

    # Get all cell names
    async def GetAllCellNames(self):
        resp = await self.cogsock.post(f"{self.session_id}/getAllCellNames", "")
        return json.dumps(resp)
        
    # Set cell name
    async def SetCellName(self, cell, name):
        await self.cogsock.post(f"{self.session_id}/setCellName", [cell, name])
        
   # Create new job
    async def CreateNewJob(self):
        await self.cogsock.post(f"{self.session_id}/createNewJob", "")   

    # Startup camera online
    async def StartupOnline(self):
       await self.cogsock.put(f"{self.session_id}/startupOnline","data")   
       
    # Get startup online status
    async def StartupOnlineStatus(self):
        resp = await self.cogsock.get(f"{self.session_id}/startupOnline")
        return json.dumps(resp)

File: JavaScript/cognex_camera.js
// An example of implemting a Cognex camera interface using the WebAPI in JavaScript.

class CogSocket {
    constructor(websocketUri) {
        this.uri = websocketUri;
        this.websocket = null;
        this.requestCounter = 0;
        
        // Maps request ID to the Promise resolve function
        this.pendingRequests = new Map(); 
        this.listeners = new Map();
        this.space = null;
        // Can be set externally (e.g., console.log)
        this.log = null; 
        
        // Event callbacks
        this.onOpen = null;
        this.onError = null;
        this.onClose = null;
    }

    Connect() {
        return new Promise((resolve, reject) => {
            try {
                this.websocket = new WebSocket(this.uri);

                this.websocket.onopen = () => {
                    if (this.onOpen) this.onOpen();
                    resolve();
                };

                this.websocket.onmessage = (event) => {
                    this._handleMessage(event.data);
                };

                this.websocket.onerror = (error) => {
                    if (this.onError) this.onError(error);
                    reject(new Error(`WebSocket Error: ${error.message || 'Unknown error'}`));
                };

                this.websocket.onclose = (event) => {
                    if (this.onClose) this.onClose(event);
                };
            } catch (e) {
                if (this.onError) this.onError(e);
                throw e;
            }
        });
    }

    _handleMessage(message) {
        if (this.log) {
            //this.log(`Received: ${message}`);
        }
        
        let msg;
        try {
            msg = JSON.parse(message);
        } catch (e) {
            // JSON Decode Error
            return;
        }
        
        const msgType = msg['$type'];
        const path = msg['path'];

        if (msgType === 'resp') {
            const reqId = msg.id;
            const pending = this.pendingRequests.get(reqId);
            
            if (pending) {
                const resolveFn = pending.resolve;
                const rejectFn = pending.reject;
                
                this.pendingRequests.delete(reqId);

                if (msg.error) {
                    const errorMsg = msg.body || String(msg.error);
                    const error = new Error(errorMsg);
                    rejectFn(error);
                } else {
                    // Attempt to cast body, assuming it should resolve to data.
                    try {
                        const body = msg.body ? JSON.parse(msg.body) : null;
                        resolveFn(body);
                    } catch(e) {
                        resolveFn(msg.body); // Resolve with raw body if JSON parsing fails
                    }
                }
            }
        } else if (msgType === 'event') {
            const body = msg.body;
            
            if (path && this.listeners.has(path)) {
                const listeners = this.listeners.get(path);
                
                // Determine arguments based on body structure
                let args = [];
                if (Array.isArray(body)) {
                    args = body;
                } else if (body !== undefined && body !== null) {
                    args = [body];
                } else {
                    args = [];
                }

                // Execute all registered listeners for this path
                for (const listener of listeners) {
                    try {
                        // In JS, we await coroutines/async functions
                        if (typeof listener.then === 'function') {
                            // Await the asynchronous listener callback
                            listener(...args);
                        } else {
                            // Execute synchronous listener
                            listener(...args);
                        }
                    } catch (e) {
                        console.error(`Error executing listener for ${path}:`, e);
                    }
                }
            }
        }
    }

    _nextRequestId() {
        this.requestCounter = (this.requestCounter + 1) % 0x7FFFFFFF;
        return this.requestCounter;
    }

    async _sendRequest(type_, path, body = null) {
        const reqId = this._nextRequestId();
        
        const promise = new Promise((resolve, reject) => {
            this.pendingRequests.set(reqId, { resolve: resolve, reject: reject });
        });

        const msg = {
            '$type': type_,
            id: reqId,
            path: path
        };
        
        if (body !== null) {
            msg.body = body;
        }
        
        const jsonMsg = JSON.stringify(msg);

        if (this.log) {
            this.log(`Send: ${jsonMsg}`);
        }

        if (!this.websocket || this.websocket.readyState !== WebSocket.OPEN) {
            const err = new Error('WebSocket is not open, cannot send request.');
            this.pendingRequests.delete(reqId);
            throw err;
        }

        this.websocket.send(jsonMsg);
        return promise;
    }

    get(path) {
        return this._sendRequest('get', path);
    }

    put(path, data) {
        return this._sendRequest('put', path, data);
    }

    post(path, data) {
        return this._sendRequest('post', path, data);
    }

    async addListener(path, listener) {
        if (!this.listeners.has(path)) {
            this.listeners.set(path, []);
            await this._sendRequest('listen', path);
        }
        this.listeners.get(path).push(listener);
    }

    async removeListener(path, listener = null) {
        if (!this.listeners.has(path)) return;
        
        const currentListeners = this.listeners.get(path);

        if (listener) {
            const index = currentListeners.indexOf(listener);
            if (index > -1) {
                currentListeners.splice(index, 1);
            }
        } else {
            // Removing all listeners
            this.listeners.delete(path);
        }

        if (!this.listeners.has(path)) {
            await this._sendRequest('unlisten', path);
        }
    }
 
    async close() {
        if (this.websocket) {
            this.websocket.close();
            this.websocket = null;
        }
    }
}

 //CognexCamera Class
class CognexCamera {

    constructor(ip, port = 80, username = 'admin', password = '') {
        this.ip = ip;
        this.port = port;
        this.username = username;
        this.password = password;
        this.cogsock = null;
        this.sessionId = null;
        this.keepAliveTimer = null;
        this.root = 'cam0/hmi';
        this.cells = 'A0:Z100';
        
        // Event subscriber lists
        this.StateChanged = [];
        this.ResultsChanged = [];
        this.LiveModeChanged = [];
        this.JobInfoChanged = [];
        this.EditorAttachedChanged = [];
        this.JobLoadingChanged = [];
        this.SettingsChanged = [];
        this.JobLoadFailed = [];
        this.JobValidationDone = [];
        this.SessionDisposed = [];

        // Initialize logging (optional)
        this.logger = (msg) => console.log(`[CogSocket] ${msg}`);
    }

    async Connect() {
        const uri = `ws://${this.ip}:${this.port}/ws`;
        this.cogsock = new CogSocket(uri);
        this.cogsock.log = this.logger;

        // Connect Socket
        await this.cogsock.Connect();
        
        // Open Session
        const sessionInfo = { cellNames: [this.cells] };
        const openSessionResponse = await this.cogsock.post(`${this.root}/openSession`, sessionInfo);
        this.sessionId = openSessionResponse && typeof openSessionResponse === 'object' && openSessionResponse.number
            ? openSessionResponse.number
            : openSessionResponse;

        if (!this.sessionId) {
            throw new Error('Unable to establish session.');
        }

        console.log(`Session ID: ${this.sessionId}`);

        // Login
        const ok = await this.cogsock.post(`${this.sessionId}/login`, [this.username, this.password, false]);
        if (ok && ok.error) {
            throw new Error("Login failed");
        }
        console.log("Login successful");

        // Add Listeners
        await this.cogsock.addListener(`${this.root}/stateChanged`, this._onStateChanged.bind(this));
        await this.cogsock.addListener(`${this.sessionId}/resultChanged`, this._onResultChanged.bind(this));
        await this.cogsock.addListener(`${this.root}/liveModeChanged`, this._onLiveModeChanged.bind(this));
        await this.cogsock.addListener(`${this.root}/jobChanged`, this._onJobChanged.bind(this));
        await this.cogsock.addListener(`${this.root}/editorAttachedChanged`, this._onEditorAttached.bind(this));
        await this.cogsock.addListener(`${this.root}/jobLoadingChanged`, this._onJobLoading_changed.bind(this));
        await this.cogsock.addListener(`${this.root}/settingsChanged`, this._onSettingsChanged.bind(this));
        await this.cogsock.addListener(`${this.root}/jobLoadFailed`, this._onJobLoadFailed_changed.bind(this));
        await this.cogsock.addListener(`${this.root}/jobValidationDone`, this._onJobValidationDone_changed.bind(this));
        await this.cogsock.addListener(`${this.root}/sessionDisposed`, this._onSessionDisposed_changed.bind(this));

        // Ready and Start Keep Alive
        await this.cogsock.post(`${this.sessionId}/ready`, null);
        this._startKeepAlive();
    }

    _startKeepAlive() {
        if (this.keepAliveTimer) {
            clearInterval(this.keepAliveTimer);
        }

        this.keepAliveTimer = setInterval(async () => {
            if (!this.sessionId || !this.cogsock) {
                return;
            }

            try {
                await this.cogsock.post(`${this.sessionId}/keepAlive`, null);
                await this.cogsock.post(`${this.sessionId}/ready`, null);
            } catch (e) {
                console.warn("Keep alive failed, possibly disconnected.", e);
                clearInterval(this.keepAliveTimer);
                this.keepAliveTimer = null;
            }
        }, 15000);
    }

    async _keepAlive() {
        while (true) {
            await new Promise(resolve => setTimeout(resolve, 15000)); // 15 seconds
            if (this.sessionId) {
                try {
                    await this.cogsock.post(`${this.sessionId}/keepAlive`, null);
                    await this.cogsock.post(`${this.sessionId}/ready`, null);
                } catch (e) {
                    console.warn("Keep alive failed, possibly disconnected.");
                    break; 
                }
            }
        }
    }

    // Events
    async _onStateChanged(...args) {
        for (const cb of this.StateChanged) {
            await cb(args);
        }
    }

    async _onResultChanged(...args) {
        this.SendReady()
        for (const cb of this.ResultsChanged ) {
            await cb(...args);
        }
    }
    
    async _onLiveModeChanged(...args) {
        for (const cb of this.LiveModeChanged) {
            await cb(...args);
        }
    }

    async _onJobChanged(...args) {
        for (const cb of this.JobInfoChanged) {
            await cb(...args);
        }
    }

    async _onEditorAttached(...args) {
        for (const cb of this.EditorAttachedChanged) {
            await cb(...args);
        }
    }

    async _onJobLoading_changed(...args) {
        for (const cb of this.JobLoadingChanged) {
            await cb(...args);
        }
    }
    
    async _onSettingsChanged(...args) {
        for (const cb of this.SettingsChanged) {
            await cb(...args);
        }
    }
    
    async _onJobLoadFailed_changed(...args) {
        for (const cb of this.JobLoadFailed) {
            await cb(...args);
        }
    }
  
    async _onJobValidationDone_changed(...args) {
        for (const cb of this.JobValidationDone) {
            await cb(...args);
        }
    }
    
    async _onSessionDisposed_changed(...args) {
        for (const cb of this.SessionDisposed) {
            await cb(...args);
        }
    }

    async Disconnect() {
        // Stop keep-alive timer
        if (this.keepAliveTimer) {
            clearInterval(this.keepAliveTimer);
            this.keepAliveTimer = null;
        }

        // Dispose session on Cognex server
        if (this.cogsock && this.sessionId) {
            try {
                console.log(`Disposing session: ${this.sessionId}`);
                await this.cogsock.post(`${this.sessionId}/dispose`, null);
            } catch (e) {
                console.error("Dispose failed:", e);
            }
        }

        // Close socket
        if (this.cogsock) {
            await this.cogsock.close();
        }

        this.sessionId = null;
        console.log("Disconnected successfully.");
    }

    // --- Camera API Methods ---

    async ManualAcquire() {
        await this.cogsock.post(`${this.sessionId}/manualTrigger`, null);
    }

    async SetLiveModeAsync(enabled) {
        await this.cogsock.put(`${this.sessionId}/liveMode`, enabled);
    }

    async ToggleOnlineOffline() {
        let resp = await this.cogsock.get(`${this.sessionId}/softOnline`);
        let currentOnline = typeof resp === 'boolean' ? resp : false;
        let newOnline = !currentOnline;
        await this.cogsock.put(`${this.sessionId}/softOnline`, newOnline);
    }

    async QueryCellResults(cell) {
        const resp = await this.cogsock.post(`${this.sessionId}/queryCellResults`, [[cell]]);
        return JSON.stringify(resp);
    }

    async GetCellExpression(cell) {
        const resp = await this.cogsock.post(`${this.sessionId}/getCellExpressions`, [cell, true]);
        return JSON.stringify(resp);
    }

    async SetCellExpression(cell, expr) {
        await this.cogsock.post(`${this.sessionId}/setCellExpression`, [cell, expr]);
    }

    async SetCellValue(cell, value) {
        await this.cogsock.post(`${this.sessionId}/setCellValue`, [cell, value]);
    }

    async ListFiles() {
        const resp = await this.cogsock.post(`${this.sessionId}/listFiles`, []);
        return JSON.stringify(resp);
    }

    async Info() {
        const resp = await this.cogsock.get(`${this.root}/info`);
        return JSON.stringify(resp);
    }

    async FindState() {
        const resp = await this.cogsock.get(`${this.root}/state`);
        return JSON.stringify(resp);
    }

    async SetSoftOnlineAsync(state) {
        await this.cogsock.put(`${this.sessionId}/softOnline`, state);
    }

    async GetJobInfo() {
        const resp = await this.cogsock.get(`${this.root}/job`);
        return JSON.stringify(resp);
    }

    async SaveJob(jobName) {
        await this.cogsock.post(`${this.sessionId}/saveJob`, jobName);
    }

    async LoadJob(jobName) {
        await this.cogsock.post(`${this.sessionId}/loadJob`, jobName);
    }

    async SendReady() {
        await this.cogsock.post(`${this.sessionId}/ready`, null);
    }

    async GetSessionIDs() {
        const resp = await this.cogsock.post(`${this.sessionId}/getSessionIDs`, null);
        return JSON.stringify(resp);
    }

    async JobValidationState() {
        const resp = await this.cogsock.get(`${this.sessionId}/jobValidationState`);
        return JSON.stringify(resp);
    }

    async SystemValidationFlag() {
        const resp = await this.cogsock.get(`${this.sessionId}/systemValidationFlag`);
        return JSON.stringify(resp);
    }
 
    async RunJobValidation() {
        await this.cogsock.post(`${this.sessionId}/runJobValidation`, null);
    }

    async CancelJobValidation() {
        await this.cogsock.post(`${this.sessionId}/cancelJobValidation`, null);
    }

    async GetKeepAliveTimeout() {
        const resp = await this.cogsock.get(`${this.root}/keepAliveTimeout`);
        return JSON.stringify(resp);
    }

    async SetKeepAliveTimeout(intervalMs) {
        const resp = await this.cogsock.put(`${this.root}/keepAliveTimeout`, intervalMs);
        return JSON.stringify(resp);
    }

    async LoadImage(fileObject) {

        const url =
            `http://${this.ip}:${this.port}/${this.sessionId}/loadImage`;

        // Convert file → base64
        const arrayBuffer = await fileObject.arrayBuffer();

        const base64 = btoa(
            String.fromCharCode(...new Uint8Array(arrayBuffer))
        );

        // Serialize string
        const jsonBody = JSON.stringify(base64);

        // Send
        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: jsonBody
        });

        if (!response.ok) {
            throw new Error(`HTTP error! Status: ${response.status}`);
        }

        return response.json();
    }
    
    async GetStartupJob() {
        const resp =await this.cogsock.get(`${this.sessionId}/startupJob`);
        return JSON.stringify(resp);
    }

    async SetStartupJob(jobName) {
        await this.cogsock.put(`${this.sessionId}/startupJob`, jobName);
    }
    
    async GetAllCellNames() {
        const resp = await this.cogsock.post(`${this.sessionId}/getAllCellNames`);
        return JSON.stringify(resp);
    }
    
    async SetCellName(cell, name) {
        await this.cogsock.post(`${this.sessionId}/setCellName`, [cell, name]);
    }

    async CreateNewJob() {
        await this.cogsock.post(`${this.sessionId}/createNewJob`, "");
    }

    async GetCellCondition(cell){
        const resp = await this.cogsock.post(`${this.sessionId}/getCellCondition`, cell);
        return JSON.stringify(resp);
    }

    async StartupOnline(state){
        await this.cogsock.put(`${this.sessionId}/startupOnline`, state);
    }

    async StartupOnlineStatus(){
        const resp = await this.cogsock.get(`${this.sessionId}/startupOnline`);
        return JSON.stringify(resp);
    }
    
}
